build-support/bin/readme_api.py

```diff
@@ -0,0 +1,264 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+"""Utilities for accessing the readme.com API.
+
+Since they don't have a Python SDK.
+
+This is not a general purpose SDK: It only covers the parts we need, namely
+Categories and Docs, and within those, only the fields we care about.
+
+Note: The readme.com API is a little confusing wrt to ids and slugs.
+ You access a Category or Doc object via its slug. But those objects reference other objects by id.
+ For example, A Doc references its category via the `category` field, which contains an id.
+ There is no easy way to go directly from an id to an object: You have to enumerate all
+ the objects. The one useful exception is that when you retrieve all the Docs in a Category,
+ you get a list of reference objects, each of which includes the id *and* the slug.
+ The fields containing ids do not have an "_id" suffix, for consistency with the over-the-wire
+ field names. So assume that any reference field contains an id, unless its name is "slug".
+"""
+
+from __future__ import annotations
+
+import dataclasses
+import json
+import logging
+from dataclasses import dataclass
+from typing import Any, Type, TypeVar, Union, cast
+
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+_API_URL_PREFIX = "https://dash.readme.io/api/v1/"
+
+
+ApiResponse = Union[dict, list, None]
+
+
+T = TypeVar("T", bound="ReadmeEntity")  # Entity type.
+F = TypeVar("F")  # Field type.
+
+
+@dataclass(frozen=True)
+class ReadmeEntity:
+    _id: str
+
+    @property
+    def id(self):
+        return self._id
+
+    @classmethod
+    def from_api_response(cls: Type[T], data: dict) -> T:
```

Comment:
fyi you can use lowercase `type` thanks to `__future__.annotations`

Reply:
Will do in a future change.